### PR TITLE
Custom fields mechanism

### PR DESCRIPTION
**Pull request by zefciu**
Added custom fields mechanism. It allows to bind custom attributes to a
content type and access them like normal attributes.

### CODE REVIEW

File: src/ralph/util/models.py [hunk 1]
```diff
@@ -109,3 +108,194 @@ def save(self, mute=False, visited=None, *args, **kwargs):
...
+    def content_type(cls):
+        return ContentType.objects.get_for_model(cls)
+
+    @property
+    def custom_attributes(cls):
+        if cls._custom_attributes is None:
+            cls.refresh_custom_attributes()
+        return cls._custom_attributes
+
+    def refresh_custom_attributes(cls):
```
Comment (kula1922) [inline src/ralph/util/models.py]: docs

File: src/ralph/util/models.py [hunk 2]
```diff
@@ -109,3 +108,194 @@ def save(self, mute=False, visited=None, *args, **kwargs):
...
+        verbose_name=_('Displayed name'),
+        max_length=255,
+    )
+    type = db.PositiveIntegerField(
+        choices=CustomAttributeTypes()
+    )
+    content_type = db.ForeignKey(ContentType)
+    required = db.BooleanField()
+
+    def load_attributes_if_necessary(self, inst):
```
Comment (kula1922) [inline src/ralph/util/models.py]: docs

File: src/ralph/util/models.py [hunk 3]
```diff
@@ -109,3 +108,194 @@ def save(self, mute=False, visited=None, *args, **kwargs):
...
+                attribute=self,
+                value=value,
+                object_id=inst.pk,
+                content_type=ContentType.objects.get_for_model(type(inst)),
+            )
+            value_object.dirty = True
+            inst.ca_values[self.type][self.pk] = value_object
+        else:
+            value_object.value = value
+            value_object.dirty = True
```
Comment (kula1922) [inline src/ralph/util/models.py]: value_object.dirty = True for each case?

File: src/ralph/util/models.py [hunk 4]
```diff
@@ -109,3 +108,194 @@ def save(self, mute=False, visited=None, *args, **kwargs):
...
+            for value in values_set:
+                value.dirty = False
+                value.to_delete = False
+                self.ca_values[type_id][value.attribute.pk] = value
+        for ca in type(self).custom_attributes:
+            value = self.ca_values[ca.type].get(ca.pk)
+            if value is not None:
+                self.__dict__[ca.name] = value.value
+            else:
+                self.__dict__[ca.name] = None
```
Comment (kula1922) [inline src/ralph/util/models.py]: self.__dict__[ca.name] = getattr(value, 'value', None)

Comment (zefciu) [inline src/ralph/util/models.py]: I really really don't like this idea, as it would exploit the fact that getattr is broken.

**Comment by vi4m** [comment]
Bye!

**Pull request closed without merging**